arrow/arrowhead.pyx

```diff
@@ -0,0 +1,149 @@
+    # considers calling __init__(), which it can skip. `self` is not yet fully
+    # constructed so *BE CAREFUL* invoking Python operations that might touch it!
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __cinit__(self, *args):
+        memset(&self.info, 0, sizeof(obsidian.Info))
+        self.info.random_state = <mersenne.MTState*> PyMem_Malloc(sizeof(mersenne.MTState))
+        if not self.info.random_state:
+            raise MemoryError()
+
+    # This will accept any type of array in C-contiguous layout that support
+    # the memoryview protocol (numpy array, array.array, Cython array, buffer
+    # interface) with the declared element type and number of dimensions.
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __init__(
+            self,
+            int random_seed,
+            int64_t[:, ::1] stoichiometry not None,
+            double[::1] rates not None,
+
+            int64_t[::1] reactants_lengths not None,
+            int64_t[::1] reactants_indexes not None,
+            int64_t[::1] reactants not None,
+            int64_t[::1] reactions not None,
+
+            int64_t[::1] dependencies_lengths not None,
+            int64_t[::1] dependencies_indexes not None,
+            int64_t[::1] dependencies not None,
+
+            int64_t[::1] substrates_lengths not None,
+            int64_t[::1] substrates_indexes not None,
+            int64_t[::1] substrates not None,
+            *args):
+        self.random_seed = random_seed
+        mersenne.seed(self.info.random_state, random_seed)
+
+        self.info.reactions_count = stoichiometry.shape[0]
+        self.info.substrates_count = stoichiometry.shape[1]
+        self.info.stoichiometry = &stoichiometry[0, 0]
+        self.info.rates = &rates[0]
+
+        self.info.reactants_lengths = &reactants_lengths[0]
+        self.info.reactants_indexes = &reactants_indexes[0]
+        self.info.reactants = &reactants[0]
+        self.info.reactions = &reactions[0]
+
+        self.info.dependencies_lengths = &dependencies_lengths[0]
+        self.info.dependencies_indexes = &dependencies_indexes[0]
+        self.info.dependencies = &dependencies[0]
+
+        self.info.substrates_lengths = &substrates_lengths[0]
+        self.info.substrates_indexes = &substrates_indexes[0]
+        self.info.substrates = &substrates[0]
+
+        # TODO(jerry): Check array sizes? E.g. rates.shape[0] >= reactions_count
+
+        self.refs = (  # hold refs to these memoryviews while we hold their data ptrs
+            stoichiometry, rates,
+            reactants_lengths, reactants_indexes, reactants, reactions,
+            dependencies_lengths, dependencies_indexes, dependencies,
+            substrates_lengths, substrates_indexes, substrates)
+
+    # `self` might no longer be a valid Python object so *BE CAREFUL* invoking
+    # Python operations that might touch it!
+    def __dealloc__(self):
```

Comment:
Under what conditions would `self` no longer be a valid python object where we could access it? Is this something we can make explicit in the calling code? Would like to ensure this mistake is not able to happen, rather than depend on callers to know this, but not sure how to accomplish that.

Reply:
`__cinit__()` and `__dealloc__()` are hooks where a Cython extension class can do its C-level allocation and deallocation.

`__cinit__()` is guaranteed to be called exactly once. It can do additional initialization beyond C allocation but object construction hasn't finished building a Python object yet, so beware.

In contrast, `__init__()` is usually called once but it could be skipped; maybe even called more than once?

It's not possible to write code that calls `__cinit__()` or `__dealloc__()`. I tested that from Python just now and got an `AttributeError: 'arrow.arrowhead.Arrowhead' object has no attribute '__calloc__'`. `__cinit__()` cannot even call its superclass `__cinit__()` method. I didn't test Cython code trying to explicitly call a `__cinit__()`.

BTW a Cython function can be declared with `def`, `cdef`, or `cpdef`. `def` gives it Python linkage and the args and result must be Python objects. If it declares arg types, the function preamble code checks all that. `cdef` gives it C linkage, the args and result can be Python and C objects, and it cannot be called from Python code. `cpdef` creates a `cdef` function and also a Python wrapper function.

`__cinit__()` and `__dealloc__()` don't fit that rule: You must declare them with `def` even though they only have C linkage and they aren't even object attributes. But they fit into another rule: All the "dunder" special methods of extension types must be declared with `def`, not `cdef`, and Python uses special calling conventions to invoke them.

There is no `__del__()` method for extension types.

BTW there are some optional features for `cdef class` that `Arrowhead` doesn't use. One can declare an attribute `cdef dict __dict__` to enable dynamic attributes. Otherwise such instances have no `__dict__`. One can declare an attribute `public` to generate get/set property accessors or `readonly` to generate a getter. There's also the more general property mechanism. Properties would be more natural than the `reactions_count()` and `substrates_count()` methods, but I didn't want to change the API.